Event Bus monitoring metrics proposal (#264)
kyma-project/community: sig-and-wg/sig-core/proposals/event-bus-monitoring-metrics.md

# Event Bus Monitoring Metrics Proposal

> 

Reviewer: Please add a description of what problem we are trying to solve here?
e.g. `The metrics aim to provide support with troubleshooting when there are issues with event delivery. As a developer, I should be able to look at the metrics and figure out why events are not delivered for my lambda/serverless.`

Author: I don't think the metrics docs should look like that because of:
- Metrics documentation usually is a plain description of what each metric describe, not the scenarios it might help with like https://docs.datadoghq.com/integrations/kafka/
- The troubleshooting scenarios can be ina different document format, each scenario and how to tackle and what metrics to look into
- Troubleshooting some scenarios would require looking into more that a metric chart and deduce what's wrong.
- Anomaly in any metric can be indicator for a problem.  
I'd suggest to have the metrics doc as a key-value (name-desc) table and move the troubelshooting scenarios and how the metrics can help to the troubleshooting guide.

Reviewer: This is a proposal. It is better to provide a context and description. And here the intention is to discover those metrics that can support troubleshooting. Just providing a plain list is not what will help.

Author: Added the motivation per metric

Author: Done

> 

Reviewer: Shouldn't we consider metrics to figure out the mismatch between Kyma subscriptions and whatever backing resources required to ensure event delivery?
There could be other scenarios where things can go wrong and Kyma subscriptions created might not be functional. Please consider all those scenarios and propose metrics for them.

## Event Bus

> - **Throughput**

Reviewer: Please assess if the metrics such as throughput are able to serve troubleshooting purpose with event delivery. Please add some details about how they can help. In case they are not in the scope of this overview, please remove them. In general, this comment applies to all the metrics

Author: I wouldn't go for removing the performance metrics as long as we are looking into system health monitoring generally. Imagine the operation is troubleshooting why lambda is not delivered for a few minutes, this might be because of a problem/miss-config or just because of the latency in the queue.
Performance metrics are important in the system troubleshooting generally and specifically in some cases.
I'd also suggest prioritizing the implementation of the proposed metrics rather than removing them.

Reviewer: Again, please provide details of why and how each metric can support troubleshooting.

Reviewer: Since this is a proposal, we need to have proper reasoning `why` such and such metric would make sense. The intent of this task is to aid the planning/refinement for the task to do the actual implementation. Just having a plain list is not sufficient for such a purpose.

Reviewer: Interesting is that you are already providing a reasonable description in the comments. It would be nice to put this in the document.

  
  Number of all events delivered that were successfully received by publish-app per minute

- **event propagation time 99 percentile**

  event propagation time 99 percentile

- **lag**

  number of events in queue

- **subscriptions**

  Number of subscriptions

- **subscriptions per {source ID &| event type}**

  Number of subscriptions per {source ID &| event type}

- **event activations**

  Number of event activations

- **consumers**

  Number of consumers
  

> - **errors per {app &| tag &| ...}**

Reviewer: What kind of errors?

Author: Golang errors like for example Internal Error during JSON marshaling, maybe this should be labeled with actor/app and any queriable other tags.
Again this should be the responsibility of Error reporting tools, but just to give some insights to the operation team that there is something wrong happens.

# Event Bus Monitoring Metrics Proposal

## Event Bus Metrics

|Metric Name |Description |Motivation |
|------------|------------|-----------|
|Throughput |Number of all events delivered that were successfully received by publish-app per minute | Gives insight about the whole system performance|

> |event propagation time 99 percentile |event propagation time 99 percentile | Gives more accurate insight about the whole system performance and events delivery|

Reviewer 2: ```suggestion
|event propagation time 99 percentile |event propagation time 99 percentile | Gives more accurate insights on the whole system performance and Event delivery|
```

> |lag | number of events in queue | Helps in diagnosing system performance as well as troubleshooting events not delivered cases|

Reviewer 3: In case of NATSS, this is gonna be per NATSS subject?

|kyma subscriptions per {namespace &\| source ID &\| event type &\| event type version &\| ready &\| endpoint} |Number of Kyma subscriptions per {namespace &\| source ID &\| event type &\| event type version &\| ready &\| endpoint} | Helps in events not delivered cases and insights about the subscriptions distribution over multiple criteria|
|event activations per {namespace &\| source ID} | Number of event activations per {namespace &\| source ID} | Can help in troubleshooting events not delivered cases and filtering the activated subscriptions based on different criteria|
|knative subscriptions per {namespace &\| channel name &\| channel ready &\| subscriber &\| ready}   |Number of knative subscriptions per {namespace &\| channel name &\| channel ready &\| subscriber &\| ready}  | Comparing this metric to the number of Kyma subscriptions can help in troubleshooting events not delivered cases and system integrity|

> |knative channels per {namespace &\| subscriber name &\| subscriber URI &\| source ID &\| event type &\| event type version} |Number of knative channels per {namespace &\| subscriber name &\| subscriber URI &\| source ID &\| event type &\| event type version} | Helps in troubleshooting events not delivered cases and system integrity by filtering the created channels by different criteria and comparing to the correspondent knative subscriptions|

Reviewer 2: ```suggestion
|knative channels per {namespace &\| subscriber name &\| subscriber URI &\| source ID &\| event type &\| event type version} |The number of knative channels per {namespace &\| subscriber name &\| subscriber URI &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues by filtering created channels by different criteria and comparing to the corresponding knative Subscriptions|
```

> |knative channels per {namespace &\| subscriber name &\| subscriber URI &\| source ID &\| event type &\| event type version} |The number of knative channels per {namespace &\| subscriber name &\| subscriber URI &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues by filtering created channels by different criteria and comparing to the corresponding knative Subscriptions|

Reviewer 2: ```suggestion
|knative channels per {namespace &\| subscriber name &\| subscriber URI &\| source ID &\| event type &\| event type version} |The number of knative channels per {namespace &\| subscriber name &\| subscriber URI &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues by filtering created channels by different criteria and comparing to the corresponding knative Subscriptions.|
```

> |consumers | Number of consumers | Helps in troubleshooting events not delivered cases and system integrity|

Reviewer 3: By consumers you mean subscribers?

Author: yes

> |consumers | Number of consumers | Helps in troubleshooting events not delivered cases and system integrity|

Reviewer 2: ```suggestion
|consumers | The number of consumers | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues|
```

> |consumers | The number of consumers | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues|

Reviewer 2: ```suggestion
|consumers | The number of consumers | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues.|
```

> |errors per {namespace &\| app &\| tag &\| ...} | The number of errors per {namespace &\| app &\| tag &\| ...}   | Helps in troubleshooting system failures, unsuccessful Event deliveries, and system integrity issues. Helps to fix system issues|

Reviewer 2: ```suggestion
|errors per {namespace &\| app &\| tag &\| ...} | The number of errors per {namespace &\| app &\| tag &\| ...}   | Helps in troubleshooting system failures, unsuccessful Event deliveries, and system integrity issues. Helps fixing system issues.|
```

> |middleware availability | Health monitor of middleware like NATS Streaming | Helps in troubleshooting system failures cases, troubleshooting events not delivered cases and system integrity|

Reviewer 2: ```suggestion
|middleware availability | The health monitor of middleware like NATS Streaming | Helps in troubleshooting system failures cases, unsuccessful Event deliveries, and system integrity issues|
```

> |middleware availability | The health monitor of middleware like NATS Streaming | Helps in troubleshooting system failures cases, unsuccessful Event deliveries, and system integrity issues|

Reviewer 2: ```suggestion
|middleware availability | The health monitor of middleware like NATS Streaming | Facilitates troubleshooting system failures cases, unsuccessful Event deliveries, and system integrity issues.|
```

## Publish App Metrics

|Metric Name |Description |Motivation |
|------------|------------|-----------|

> |events in |Total Number of received events | Gives insights about system load|

Reviewer 2: ```suggestion
|events in total |The number of received Events | Provides insights on the system load|
```

`Total` should be included in the metric name 🤔

> |events in total |The number of received Events | Provides insights on the system load|

Reviewer 2: ```suggestion
|events in total |The number of received Events | Provides data on the system load.|
```

> |events in per {namespace &\| source ID &\| event type &\| event type version} |Number of events requests per {namespace &\| source ID &\| event type &\| event type version} | Filtering received events by different criteria combination can help with giving a better insight about the system load as well as troubleshooting events not delivered cases and system integrity|

Reviewer 2: ```suggestion
|events in total per {namespace &\| source ID &\| event type &\| event type version} |The number of Event requests per {namespace &\| source ID &\| event type &\| event type version} | Filtering received Events by different criteria combination provides a better insight on the system load as well as facilitates troubleshooting unsuccessful Event deliveries and system integrity issues|
```

I believe the metric name lacked `total` as in `Events in total` :)

> |events in total per {namespace &\| source ID &\| event type &\| event type version} |The number of Event requests per {namespace &\| source ID &\| event type &\| event type version} | Filtering received Events by different criteria combination provides a better insight on the system load as well as facilitates troubleshooting unsuccessful Event deliveries and system integrity issues|

Reviewer 2: ```suggestion
|events in total per {namespace &\| source ID &\| event type &\| event type version} |The number of Event requests per {namespace &\| source ID &\| event type &\| event type version} | Filtering received Events by different criteria combination provides a better insight into the system load as well as facilitates troubleshooting unsuccessful Event deliveries and system integrity issues.|
```

> |succeeded events per {namespace &\| source ID &\| event type &\| event type version} | Number of succeeded events per {namespace &\| source ID &\| event type &\| event type version} | Helps in troubleshooting events not delivered cases and system integrity as well as being a system health indicator|

Reviewer 2: ```suggestion
|succeeded events per {namespace &\| source ID &\| event type &\| event type version} | The number of successful Events per {namespace &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues as well as acts as the system health indicator|
```

> |succeeded events per {namespace &\| source ID &\| event type &\| event type version} | The number of successful Events per {namespace &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues as well as acts as the system health indicator|

Reviewer 2: ```suggestion
|succeeded events per {namespace &\| source ID &\| event type &\| event type version} | The number of successful Events per {namespace &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues. It also acts as the system health indicator.|
```

> |ignored events per {namespace &\| source ID &\| event type &\| event type version} | Number of ignored events per {namespace &\| source ID &\| event type &\| event type version} | Helps in troubleshooting events not delivered cases and system integrity as well as being a system health indicator|

Reviewer 2: ```suggestion
|ignored events per {namespace &\| source ID &\| event type &\| event type version} | The number of ignored events per {namespace &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues as well as acts as the system health indicator|
```

> |ignored events per {namespace &\| source ID &\| event type &\| event type version} | The number of ignored events per {namespace &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues as well as acts as the system health indicator|

Reviewer 2: ```suggestion
|ignored events per {namespace &\| source ID &\| event type &\| event type version} | The number of ignored events per {namespace &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues. It also acts as the system health indicator.|
```

> |failed events per {namespace &\| source ID &\| event type &\| event type version} | Number of failed events per {namespace &\| source ID &\| event type &\| event type version} | Helps in troubleshooting events not delivered cases and system integrity as well as being a system health indicator|

Reviewer 2: ```suggestion
|failed events per {namespace &\| source ID &\| event type &\| event type version} | Number of failed events per {namespace &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues as well as acts as the system health indicator|
```

> |failed events per {namespace &\| source ID &\| event type &\| event type version} | Number of failed events per {namespace &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues as well as acts as the system health indicator|

Reviewer 2: ```suggestion
|failed events per {namespace &\| source ID &\| event type &\| event type version} | Number of failed events per {namespace &\| source ID &\| event type &\| event type version} | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues. It also acts as the system health indicator.|
```

> |latency 99 percentile | Latency 99 percentile of published event request | High latency maybe a reason for not delivered events in some cases and it gives an insight about the system health and performance

Reviewer 2: ```suggestion
|latency 99 percentile | Latency 99 percentile of published event request | High latency may be a reason for unsuccessful Event deliveries in some cases so it gives more insight on the system health and performance
```

> |latency 99 percentile | Latency 99 percentile of published event request | High latency may be a reason for unsuccessful Event deliveries in some cases so it gives more insight on the system health and performance

Reviewer 2: ```suggestion
|latency 99 percentile | Latency 99 percentile of published event request | High latency may be a reason for unsuccessful Event deliveries in some cases so it provides more data on the system health and performance.
```

## Push App Metrics

|Metric Name |Description |Motivation |
|------------|------------|-----------|

> |events in | Number of events push requests | Comparing this metric to the received events by the publish app metric can help in troubleshooting events not delivered cases and system integrity|

Reviewer 2: ```suggestion
|events in total | The number of Events push requests | Comparing this metric its counterpart from the publish-app metric facilitates troubleshooting unsuccessful Event deliveries and system integrity issues|
```
Seems that there is something missing - should it be `events in total`?

> |events in total | The number of Events push requests | Comparing this metric its counterpart from the publish-app metric facilitates troubleshooting unsuccessful Event deliveries and system integrity issues|

Reviewer 2: ```suggestion
|events in total | The number of Events push requests | Comparing this metric its counterpart from the publish-app metric facilitates troubleshooting unsuccessful Event deliveries and system integrity issues.|
```

> |pushed messages | Number of pushed messages to all consumers | Giving insight about knative eventing health by monitoring the rates as well as assuring the system integrity|

Reviewer 2: ```suggestion
|pushed messages | The number of messages pushed to all consumers | Provides insight on knative eventing health by monitoring the rates as well as ensuring system integrity|
```

> |pushed messages | The number of messages pushed to all consumers | Provides insight on knative eventing health by monitoring the rates as well as ensuring system integrity|

Reviewer 2: ```suggestion
|pushed messages | The number of messages pushed to all consumers | Provides insights into knative eventing health by monitoring the rates as well as ensuring system integrity.|
```

> |latency 99 percentile to all consumers|Latency 99 percentile of pushed events to all consumers | Gives insight about the knative/events delivery system performance and can help in troubleshooting events not delivered scenarios|

Reviewer 2: ```suggestion
|latency 99 percentile to all consumers|Latency 99 percentile of pushed events to all consumers | Provides insight on the knative/Event delivery system performance and facilitates troubleshooting unsuccessful Event deliveries|
```

> |latency 99 percentile to all consumers|Latency 99 percentile of pushed events to all consumers | Provides insight on the knative/Event delivery system performance and facilitates troubleshooting unsuccessful Event deliveries|

Reviewer 2: ```suggestion
|latency 99 percentile to all consumers|Latency 99 percentile of pushed events to all consumers | Provides insight into knative/Event delivery system performance and facilitates troubleshooting unsuccessful Event deliveries.|
```

> |latency 99 percentile to a single consumer | Latency 99 percentile of pushed message to a single consumer | Assessing the system delivery performance and contributes to judging the whole system performance diagnosis|

Reviewer 2: ```suggestion
|latency 99 percentile to a single consumer | Latency 99 percentile of messages pushed to a single consumer | Assesses system delivery performance and contributes to the diagnosis of the general system performance|
```

> |failed pushed messages | Number of failed pushed messages | Help in troubleshooting events not delivered scenarios and system integrity|

Reviewer 3: I'm guessing you'll consider the delivery status code? 
https://github.com/kyma-project/kyma/blob/3c2c64319e867bdf6cbace9717c2457a9f1d0ddf/components/event-bus/internal/push/handlers/message_handler.go#L110

Reviewer 2: ```suggestion
|failed pushed messages | The number of failed pushed messages | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues|
```

> |latency 99 percentile to a single consumer | Latency 99 percentile of messages pushed to a single consumer | Assesses system delivery performance and contributes to the diagnosis of the general system performance|

Reviewer 2: ```suggestion
|latency 99 percentile to a single consumer | Latency 99 percentile of messages pushed to a single consumer | Assesses system delivery performance and contributes to the diagnosis of the general system performance.|
```

> |failed pushed messages | The number of failed pushed messages | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues|

Reviewer 2: ```suggestion
|failed pushed messages | The number of failed pushed messages | Facilitates troubleshooting unsuccessful Event deliveries and system integrity issues.|
```
